DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/ViewDelegateController.cs

```diff
@@ -12,29 +14,54 @@
     [Route("TrackingSystem/Delegates/View/{delegateId:int}")]
     public class ViewDelegateController : Controller
     {
+        private readonly ICourseDataService courseDataService;
         private readonly CustomPromptHelper customPromptHelper;
+        private readonly ICustomPromptsService customPromptsService;
         private readonly IUserDataService userDataService;
 
-        public ViewDelegateController(IUserDataService userDataService, CustomPromptHelper customPromptHelper)
+        public ViewDelegateController(
+            IUserDataService userDataService,
+            CustomPromptHelper customPromptHelper,
+            ICourseDataService courseDataService,
+            ICustomPromptsService customPromptsService
+        )
         {
             this.userDataService = userDataService;
             this.customPromptHelper = customPromptHelper;
+            this.courseDataService = courseDataService;
+            this.customPromptsService = customPromptsService;
         }
 
         public IActionResult Index(int delegateId)
         {
             var centreId = User.GetCentreId();
-            var delegateUser = userDataService.GetDelegateUserCardById(delegateId);
 
+            var delegateUser = userDataService.GetDelegateUserCardById(delegateId);
             if (delegateUser == null || delegateUser.CentreId != centreId)
             {
                 return new NotFoundResult();
             }
 
             var customFields = customPromptHelper.GetCustomFieldViewModelsForCentre(centreId, delegateUser);
-            var delegateInfo = new DelegateInfoViewModel(delegateUser, customFields);
-            var model = new ViewDelegateViewModel(delegateInfo);
+            var delegateInfoViewModel = new DelegateInfoViewModel(delegateUser, customFields);
+
+            var courseInfoViewModels = courseDataService.GetDelegateCoursesInfo(delegateId)
+                .Select(
+                    info =>
+                    {
+                        var courseCustomPromptsWithAnswers = customPromptsService.GetCustomPromptsWithAnswersForCourse(
+                            info,
+                            info.CustomisationId,
+                            centreId,
+                            includeArchived: true
+                        );
+                        var attemptStats =
+                            courseDataService.GetDelegateCourseAttemptStats(delegateId, info.CustomisationId);
+                        return new DelegateCourseInfoViewModel(info, courseCustomPromptsWithAnswers, attemptStats);

```

Comment:
I don't think all this logic should be in the controller. I think it should be in the Service level, populate all the information we need for the view into a single entity (or tuple if necessary).

i.e. a new method in CourseService GetAllCoursesForDelegate that essentially does the content of the lambda. Then this call just becomes Select(x => new DelegateCourseInfoViewModel(x))